File: llm-security-engine/app/routes/health.py
import time
import httpx
from fastapi import APIRouter
from app.config import get_settings
from app.services.ollama_client import check_ollama_connectivity
from app.models.schemas import OllamaPingResponse

router = APIRouter()
settings = get_settings()


@router.get("/health")
async def health_check():
    """
    Service health check. Reports service status and Ollama connectivity.
    Always returns 200 — Ollama unreachability is reported inside the response body.
    """
    ollama_status = await check_ollama_connectivity()

    overall = "ok" if ollama_status["reachable"] else "degraded"

    return {
        "status": overall,
        "service": "Local LLM Security Engine",
        "version": "1.0.0",
        "config": {
            "ollama_base_url": settings.OLLAMA_BASE_URL,
            "ollama_model": settings.OLLAMA_MODEL,
            "ollama_timeout_seconds": settings.OLLAMA_TIMEOUT,
            "debug": settings.DEBUG,
        },
        "ollama": ollama_status,
    }


@router.get("/debug/ping-ollama", response_model=OllamaPingResponse)
async def ping_ollama():
    """
    Development-only endpoint: lightweight Ollama connectivity probe.
    Returns reachability, configured model availability, and round-trip latency.
    """
    start = time.monotonic()
    status = await check_ollama_connectivity()
    latency_ms = round((time.monotonic() - start) * 1000, 2)

    return OllamaPingResponse(
        reachable=status["reachable"],
        base_url=settings.OLLAMA_BASE_URL,
        configured_model=settings.OLLAMA_MODEL,
        model_available=status.get("model_available", False),
        available_models=status.get("available_models", []),
        error=status.get("error"),
        latency_ms=latency_ms,
    )

File: soc-backend/src/routes/health.ts
import { Router, type IRouter, type Request } from "express";
import { HealthCheckResponse } from "@workspace/api-zod";
import { getConfig } from "../lib/config.js";
import { getLocalEngineClient } from "../providers/localSecurityEngine.js";
import { logger } from "../lib/logger.js";

const router: IRouter = Router();

/**
 * GET /api/healthz
 *
 * Lightweight liveness check. Always returns 200 with {status: "ok"}.
 * Use /api/provider-health for detailed provider connectivity status.
 */
router.get("/healthz", (_req, res) => {
  const data = HealthCheckResponse.parse({ status: "ok" });
  res.json(data);
});

/**
 * GET /api/provider-health
 *
 * Probes the configured LLM provider and returns connectivity status.
 * Never returns an error status — always 200 with reachability details.
 *
 * Response fields:
 *   provider_mode        — active mode ("local_security_engine" | "none")
 *   configured_base_url  — raw value of LOCAL_LLM_ENGINE_BASE_URL (null if unset)
 *   engine_reachable     — true if the engine responded to the health probe
 *   engine_status        — engine's self-reported status ("ok" | "degraded" | null)
 *   model_name           — Ollama model configured in the engine (null if unknown)
 *   latency_ms           — round-trip latency to engine's /health endpoint (null if not probed)
 *   auth_configured      — true if LOCAL_LLM_ENGINE_API_KEY is set
 *   engine_error         — error message if unreachable (null if healthy)
 */
router.get("/provider-health", async (req: Request, res) => {
  const requestId = (req.headers["x-request-id"] as string | undefined) ?? crypto.randomUUID();
  const config = getConfig();

  if (config.providerMode === "none") {
    logger.info({ request_id: requestId, provider_mode: "none" }, "provider_health_no_config");
    res.json({
      provider_mode: "none",
      configured_base_url: null,
      engine_reachable: false,
      engine_status: null,
      model_name: null,
      latency_ms: null,
      auth_configured: false,
      engine_error: "No provider configured. Set LOCAL_LLM_ENGINE_BASE_URL.",
    });
    return;
  }

  const { baseUrl, apiKey, timeoutMs } = config.localEngine;
  const client = getLocalEngineClient(baseUrl!, apiKey, timeoutMs);

  logger.debug({ request_id: requestId, engine_url: baseUrl }, "provider_health_probe");

  const health = await client.checkHealth();

  logger.info(
    {
      request_id: requestId,
      engine_reachable: health.engine_reachable,
      engine_status: health.engine_status,
      engine_model: health.engine_model,
      latency_ms: health.latency_ms,
    },
    "provider_health_result",
  );

  res.json({
    provider_mode: config.providerMode,
    configured_base_url: baseUrl,
    engine_reachable: health.engine_reachable,
    engine_status: health.engine_status,
    model_name: health.engine_model,
    latency_ms: health.latency_ms,
    auth_configured: health.auth_configured,
    engine_error: health.engine_error,
  });
});

export default router;
